src/app/static/src/app/components/genericChart/GenericChartTable.vue

```diff
@@ -72,30 +69,80 @@
+               // to save the hierarchy string for each option. This traverses the entire hierarchy once only, which is
+               // quicker than calling findPath on each row value.
+               const addPathsToDict = (option: NestedFilterOption, parentPath: string,  dict: Dict<string>) => {
+                   //Add the path for this item...
+                   dict[option.id] = parentPath;
+                   //...then adds paths for any children
+                   const path = parentPath + (parentPath.length ? "/" : "") + option.label;
+                   if (option.children) {
+                       option.children.forEach((child) => addPathsToDict(child as NestedFilterOption, path, dict));
+                   }
+               };
+
+               const result = {} as Dict<Dict<string>>;
+               this.columnsWithHierarchy.forEach((columnConfig) => {
+                   const columnId = columnConfig.data.hierarchyColumn!;
+                   const pathsDict = {};
+                   const column = this.columns.find((column) => column.id == columnId)!;
+                   // Deal with both array or object as top-level values type
+                   const values = Array.isArray(column.values) ? column.values : [column.values];
+                   values.forEach((value) => addPathsToDict(value, "", pathsDict));
+                   result[columnId] = pathsDict;
+               });
+               return result;
+            },
             labelledData() {
                 // Build a dictionary of the columns defined in the fetched dataset, which include all values and
                 // associated labels
                 const columnsDict = this.columns.reduce((dict, column) => ({...dict, [column.id]: column}), {} as Dict<GenericChartColumn>);
-                return this.filteredData.map(row => {
+
+                // Convert each label column's filter options to a dictionary of id to labels so we do not need to find values'
+                // labels in the options array per row
+                const configLabelColumns = this.tableConfig.columns.filter(column => column.data.labelColumn);
+                const columnValueLabels = configLabelColumns.reduce((dict, columnConfig) => {
+                    const key = columnConfig.data.labelColumn!;
+                    // If key is not already in dictionary, and options are an array (not always the case for hierarchies)
+                    if (!dict[key] && columnsDict[key].values.reduce) {
+                        const filterOptions = columnsDict[key].values;
+                        const labelDict = filterOptions.reduce((dict, option) => ({
+                            ...dict,
+                            [option.id]: option.label
+                        }), {} as Dict<string>);
+
+                        return {
+                            ...dict,
```

Comment:
Just clocked that we clone every row in the inner loop (line 126) which will be slowing things down quite a bit but I can't really think of any way around that. Might be worth looking at the chart component again and just checking for any other instances of this slight anti-pattern.

Reply:
Yeah, we can't really edit the original data which is in the state. The chart does not do this per row clone, as far as I can tell.

Reply:
Actually, this isn't true! The chart *does* copy and modify rows, in order to add page numbers. But we don't want the chart and table to be sharing the same copied rows since the chart may need the 'unfriendly' values. But I think I'll try making the table not copy the whole row, but only the columns it needs.